types/valuerenderer/valuerenderer.go

```diff
@@ -0,0 +1,206 @@
+
+		sb.WriteString(formatedAmount)
+		sb.WriteString(metadata.Display)
+
+	default:
+		panic("type is invalid")
+	}
+
+	return sb.String(), nil
+}
+
+// TODO address the cass where denom starts with "u"
+func convertToBaseDenom(denom string) string {
+	switch {
+	// e.g. uregen => uregen
+	case strings.HasPrefix(denom, "u"):
+		return denom
+	// e.g. mregen => uregen
+	case strings.HasPrefix(denom, "m"):
+		return "u" + denom[1:]
+	// has no prefix regen => uregen
+	default:
+		return "u" + denom
+	}
+}
+
+// computeExponentSubtraction iterates over metadata.DenomUnits and computes the subtraction of exponents
+func computeExponentSubtraction(denom string, metadata banktypes.Metadata) float64 {
+	var coinExp, displayExp int64
+	for _, denomUnit := range metadata.DenomUnits {
+		if denomUnit.Denom == denom {
+			coinExp = int64(denomUnit.Exponent)
+		}
+
+		if denomUnit.Denom == metadata.Display {
+			displayExp = int64(denomUnit.Exponent)
+		}
+	}
+
+	return float64(coinExp - displayExp)
+}
+
+// removeTrailingZeroes removes trailing zeroes from a string
+func removeTrailingZeroes(str string) string {
+	index := len(str)-1
+	for i := len(str) - 1; i > 0; i-- {
+		if rune(str[i]) == rune('0') {
+			continue
+		} 
+		
+		index = i
+		break
+	}
+
+	return str[:index+1]
+}
+
+// countTrailingZeroes counts the amount of trailing zeroes in a string
+func countTrailingZeroes(str string) int {
+	counter := 0
+	for i := len(str) - 1; i > 0; i-- {
+		if rune(str[i]) == rune('0') {
+			counter++
+		} else {
+			break
+		}
+	}
+	return counter
+}
+
+// ComputeAmount calculates an amount to produce formated output
+func (dvr DefaultValueRenderer) ComputeAmount(amount int64, expSub float64) string {
+
+	switch {
+	// negative , convert mregen to regen less zeroes 23 => 0,023, expSub -3
+	case math.Signbit(expSub):
+
+		stringValue := strconv.FormatInt(amount, 10)
+		count := countTrailingZeroes(stringValue)
+		if count >= int(math.Abs(expSub)) {
+			// case 1 if number of zeroes >= Abs(expSub)  23000, -3 => 23 (int64)
+			x := amount / int64(math.Pow(10, math.Abs(expSub)))
+			return humanize.Comma(x)
+		} else {
+			// case 2 number of trailing zeroes < abs(expSub)  23, -3,=> 0.023(float64)
+			x := float64(float64(amount) / math.Pow(10, math.Abs(expSub)))
+			return humanize.Ftoa(x)
+		}
+	// positive, e.g.convert mregen to uregen
+	case !math.Signbit(expSub):
+		x := amount * int64(math.Pow(10, expSub))
+		return humanize.Comma(x)
+	// == 0, convert regen to regen, amount does not change
+	default:
+		return humanize.Comma(amount)
+	}
+}
+
+// Parse parses a string and takes a decision whether to convert it into Coin or Uint
+func (dvr DefaultValueRenderer) Parse(ctx context.Context, s string) (interface{}, error) {
+	if s == "" {
+		return nil, errors.New("unable to parse empty string")
+	}
+	// remove all commas
+	str := strings.ReplaceAll(s, ",", "")
+	re := regexp.MustCompile(`(\d+)(\w+)`)
```

Comment:
we should compile the regexp once, and store it as a module private variable.

Reply:
I would rather make it global since `valuerenderer_test` package imports that.